### PR TITLE
Enable frequency compensation on SX1280

### PR DESCRIPTION
**Pull request by FrankPetrilli**
See page 131 of the SX1280 datasheet: "After SetModulationParams... In all cases 0x1 must be written to the Frequency Error Compensation mode register 0x093"

Still testing this out, but there's a chance this could reduce the effect of temperature skew on certain devices.

### CODE REVIEW

**Comment by SunjunKim** [comment]
This doesn’t change anything, on all my tests.
Anyway it could be good to add this line, following the datasheet.

**Comment by CapnBry** [comment]
@FrankPetrilli Mind if I push some changes to this PR which enables our frequency compensation of the FEI is available? It works but we can't use it because apparently FEI only works if the explicit header is enabled.

**Comment by FrankPetrilli** [comment]
Sure! Let me know if there's anything you'd like me to do.

**Comment by CapnBry** [comment]
I've pushed my changes that enable the Frequency Correction on Team2.4. Note that it doesn't work unless you turn on VARIABLE_LENGTH packets (with the header) and then you need to extend the period interval in common.cpp if you want to stay in sync.

With DEBUG_LOG set, you can see the FreqCorr when the RX disconnects with the `fo=X fc=Y`, where `fc` is the correction applied, in reg units (and `fo` is the timing offset).

**Comment by CapnBry** [comment]
Oh and I wanted to apologize again for sitting on my changes for so long. We had a whole discussion in #dev, and then you said I could push changes, but then completely forgot about it. Too many PRs happening!

**Comment by pkendall64** [comment]
> I've pushed my changes that enable the Frequency Correction on Team2.4. Note that it doesn't work unless you turn on VARIABLE_LENGTH packets (with the header) and then you need to extend the period interval in common.cpp if you want to stay in sync.
> 
> With DEBUG_LOG set, you can see the FreqCorr when the RX disconnects with the `fo=X fc=Y`, where `fc` is the correction applied, in reg units (and `fo` is the timing offset).

Rather that adjusting the interval, could we just reduce the preamble by 4 to keep things close to the existing TOA?

**Comment by CapnBry** [comment]
> Rather that adjusting the interval, could we just reduce the preamble by 4 to keep things close to the existing TOA?

I don't think we want to do that, since reducing the preamble length reduces receiver sensitivity. It would be a tradeoff where adding the header increases sensitivity due to better tuning, but reduces sensitivity due to reduced preamble. I'd much prefer a way where we can keep the same preamble but somehow get FEI information.

I considered making one of the packet rates "calibration", where the header is added and the timing is stretched so the RX can figure out what the calibration offset needs to be and store it, but that is tied to a specific TX. It would be bad if with one TX the RX needed +400 and for another it needed -400 so it can't connect at all with the +400 stored. I don't think one time calibration will work for this reason.

I don't really have a good solution, and with our 500Hz packet probably growing at least 1 byte, we really don't have the TOA to be able to add the header, even with 4 symbol preamble. It also is a waste of normal bandwidth to use explicit header, considering we know our packet size in advance so the TOA is wasted to send it.

**Comment by CapnBry** [comment]
I've updated this against master since I think we should have this in mainline to have the _ability_ to do FreqCorrection even if our current timing can not use it.

I've also altered `HandleFreqCorr` so that it doesn't reset when it hits the min/max value. This generally leads to it just going from 0 to Min/Max and resetting over and over and over, which makes the signal better then worse and repeats real quick. Instead if it hits the end it just stops adjusting in that direction.

**Comment by andlier** [comment]
@CapnBry how about using a fraction of the telemetry slots for sending packets with explicit header for freq-correction? These packets can be with less payload, so could meet timing?

**Comment by CapnBry** [comment]
> @CapnBry how about using a fraction of the telemetry slots for sending packets with explicit header for freq-correction? These packets can be with less payload, so could meet timing?

The slots are all the same duration, be they going RX->TX or TX->RX. They won't fit.

Enabling it seems to be a moot point though, since sunjun has done logging and found that as long as the signal is inside the 20% tolerance specified in the datasheet (+/- 800 of our FREQ_STEP), there is zero increase in RSSI or SNR. If it is outside that range then there is no packet reception at all and therefore no FEI available to determine which way the offset needs to move.

**Comment by andlier** [comment]
> The slots are all the same duration, be they going RX->TX or TX->RX. They won't fit. 

I get same on-air-time in the calculator if I reduce payload from 8 to 5 bytes and enable explicit header, why would it not fit?

> Enabling it seems to be a moot point though, since sunjun has done logging and found that as long as the signal is inside 
the 20% tolerance specified in the datasheet (+/- 800 of our FREQ_STEP), there is zero increase in RSSI or SNR. If it is outside that range then there is no packet reception at all and therefore no FEI available to determine which way the offset needs to move.

I think I had a similar discussion on discord with @JBKingdon could we just try a few different offsets until there is a connection? and in any case, frequency correction would help to keep a stable link if temperatures changes during flight?

**Comment by CapnBry** [comment]
Oh yeah I guess you could reduce the packet size for some telemetry packets then do freqcorr on the TX side (it is currently only done on the RX side), but again there's no point to doing it. If we're getting packets, that's already close enough that there's no reason to adjust the frequency any further.

We can't really shift the frequency around either to try to get a connection, because the connection sync process already is really slow and only has a couple opportunities to connect per cycle. If we shift all over the place it would reduce our chance of connecting even further making it even more problematic than it already is.

Ultimately though, the question is why would we do any of this. If a manufacturer creates a product that is so far out of spec that we can't connect, they need to fix that product, not make us bend over backwards and spend dozens of hours of developer time to create a workaround. We also do not want to add a "calibration procedure" as a step in everyone's build process. Nor do we want "Oh your RX can't connect? Well you probably need to calibrate it" to be one of the troubleshooting steps. Receivers should be within the generous 20% frequency tolerance, not worked around in software.

**Comment by andlier** [comment]
@CapnBry  yeah, I see those points, I just saw some chatter about this on discord and here lately, and found it an interesting topic. 
A different idea: have some kind of freq-offset measurement debug-mode for checking how much off the clocks are between tx and rx? Maybe just as a separate compile option or even standalone project... that would've been nice for the DIY-ers and for the advanced users to verify and give peace of mind that their existing hw would work reliably, especially since FLRC-mode require tighter tolerances?

File: src/src/rx_main.cpp [hunk 1]
```diff
@@ -538,16 +543,14 @@ void ICACHE_RAM_ATTR HWtimerCallbackTock()
     bool didFHSS = HandleFHSS();
     bool tlmSent = HandleSendTelemetryResponse();
 
-    #if defined(RADIO_SX127X)
-    if (!didFHSS && !tlmSent && LQCalc.currentIsSet())
+    if (!didFHSS && !tlmSent && LQCalc.currentIsSet() && Radio.FrequencyErrorAvailable())
     {
         HandleFreqCorr(Radio.GetFrequencyErrorbool());      // Adjusts FreqCorrection for RX freq offset
+    #if !defined(Regulatory_Domain_ISM_2400)
```
Comment (cruwaller) [inline src/src/rx_main.cpp:549]: this should be `#if defined(RADIO_SX127X)`

File: src/src/rx_main.cpp [hunk 2]
```diff
@@ -244,7 +244,11 @@ void SetRFLinkRate(uint8_t index) // Set speed of RF link
     expresslrs_rf_pref_params_s *const RFperf = get_elrs_RFperfParams(index);
     bool invertIQ = UID[5] & 0x01;
 
-    hwTimer.updateInterval(ModParams->interval);
+  uint32_t interval = ModParams->interval;
```
Comment (cruwaller) [inline src/src/rx_main.cpp:247]: Indentation

**Comment by SunjunKim** [comment]
The build failed on configurator with FM30_RX_MINI target:
```
fatal: no tag exactly matches '7d89c22d67cea14213d6d948e3d5322554ba0b90'
.pio\libdeps\FM30_RX_MINI_via_BetaflightPassthrough\extEEPROM\extEEPROM.cpp: In member function 'byte extEEPROM::read(long unsigned int, byte*, unsigned int)':
.pio\libdeps\FM30_RX_MINI_via_BetaflightPassthrough\extEEPROM\extEEPROM.cpp:185:51: warning: ISO C++ says that these are ambiguous, even though the worst conversion for the first is better than the worst conversion for the second:
  185 |         communication->requestFrom(ctrlByte, nRead);
      |                                                   ^
In file included from .pio\libdeps\FM30_RX_MINI_via_BetaflightPassthrough\extEEPROM/extEEPROM.h:62,
                 from .pio\libdeps\FM30_RX_MINI_via_BetaflightPassthrough\extEEPROM\extEEPROM.cpp:58:
C:\.platformio\packages\framework-arduinoststm32\libraries\Wire\src/Wire.h:99:13: note: candidate 1: 'uint8_t TwoWire::requestFrom(int, int)'
   99 |     uint8_t requestFrom(int, int);
      |             ^~~~~~~~~~~
C:\.platformio\packages\framework-arduinoststm32\libraries\Wire\src/Wire.h:95:13: note: candidate 2: 'uint8_t TwoWire::requestFrom(uint8_t, uint8_t)'
   95 |     uint8_t requestFrom(uint8_t, uint8_t);
      |             ^~~~~~~~~~~
lib\LUA\rx_devLUA.cpp:52:31: warning: 'luaRxWebUpdate' defined but not used [-Wunused-variable]
   52 | static struct luaItem_command luaRxWebUpdate = {
      |                               ^~~~~~~~~~~~~~
c:/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot open map file C:/Users/����/AppData/Roaming/ExpressLRS Configurator/firmwares/github/ExpressLRS/src/.pio/build/FM30_RX_MINI_via_BetaflightPassthrough/firmware.map: Invalid argument
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\FM30_RX_MINI_via_BetaflightPassthrough\firmware.elf] Error 1
========================= [FAILED] Took 39.11 seconds =========================
Environment                             Status    Duration
--------------------------------------  --------  ------------
FM30_RX_MINI_via_BetaflightPassthrough  FAILED    00:00:39.111
==================== 1 failed, 0 succeeded in 00:00:39.111 ====================
Error: spawn "C:\.platformio\penv\Scripts\platformio.exe" ENOENT
```

**Comment by SunjunKim** [comment]
I found that with `DEBUG_FREQ_CORRECTION` `disabled` SNR is not updated...

I updated through the configurator PR tab, with normal options I've used.
TX: Namimno OLED v2
RX: SIYI Fr mini, HM EP2
Let me test more later.

File: src/src/rx_main.cpp (hunk 1 above)
Comment (CapnBry) [reply src/src/rx_main.cpp:549]: Oh snap how did that get in there? Nice catch, fixed.

**Comment by SunjunKim** [comment]
When I flash a Namimno diversity RX with the following user define, it gets into a bootloop with some errors:

```
-DMY_BINDING_PHRASE="-----"
-DRegulatory_Domain_ISM_2400
-DLOCK_ON_FIRST_CONNECTION
-DUSE_DIVERSITY
-DUART_INVERTED
-DAUTO_WIFI_ON_INTERVAL=60
-DDEBUG_LOG
-DDEBUG_CRSF_NO_OUTPUT
-DDEBUG_FREQ_CORRECTION
```

```
Hal Init
Use TX pin: 9
PLATFORM_ESP8266
SX1280 Reset
SX1280 Ready!
SX1280 Begin
Read Vers: 43447
SetPower: 31
Config FLRC
About to start CRSF task...

--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Exception (29):
epc1=0x402027ba epc2=0x00000000 epc3=0x00000000 excvaddr=0x08f00000 depc=0x00000    000

>>>stack>>>

ctx: cont
sp: 3ffffdb0 end: 3fffffc0 offset: 0190
3fffff40:  402065d4 00000002 00000000 402027ac
3fffff50:  3fff1904 0000010c ffffffff 402024f9
3fffff60:  3fff1904 3fff00d8 3ffeff24 3ffe85d4
3fffff70:  3fff1904 3fff00d8 3ffeff24 4020140b
3fffff80:  00000001 feefeffe feefeffe feefeffe
3fffff90:  feefeffe feefeffe feefeffe 3fff1a34
3fffffa0:  3fffdad0 00000000 3fff1a20 402089d4
3fffffb0:  feefeffe feefeffe 3ffe8810 40102dc9
<<<stack<<<

--------------- CUT HERE FOR EXCEPTION DECODER ---------------
```

**Comment by SunjunKim** [comment]
I think it doesn't spit out debug messages on STM32 RXes? (e.g., Siyi FR mini)?

**Comment by CapnBry** [comment]
> I think it doesn't spit out debug messages on STM32 RXes? (e.g., Siyi FR mini)? Could you check whether I did something wrong or it's a general problem?

```
Config LoRa
2000
tentative conn
got conn
Timer locked
lost conn fc=-52 fo=7
Config LoRa
```

Seems to work for me?

File: src/lib/SX1280Driver/SX1280.cpp [hunk 3]
```diff
@@ -229,6 +234,8 @@ void SX1280Driver::ConfigModParamsLoRa(uint8_t bw, uint8_t sf, uint8_t cr)
     default:
         hal.WriteRegister(0x925, 0x32); // for SF9, SF10, SF11, SF12
     }
+    // Enable frequency compensation
+    hal.WriteRegister(0x93c, 0x1);
```
Comment (JyeSmith) [inline src/lib/SX1280Driver/SX1280.cpp:238]: Should 0x93c be added to SX1280_Regs.h?

**Review by SunjunKim — APPROVED** [review]
I've been tested this and it's working OK so far

**Review by pkendall64 — APPROVED** [review]

**Pull request merged**